torch/jit.py

```diff
@@ -1,26 +1,375 @@
+        # with methods
+        @raw_trace(nderivs=self.nderivs)
+        def traced_inner(in_vars, in_struct):
+            # NB: Uncommenting this line should be equivalent
+            # args, _params = _unflatten(in_vars, in_struct)
+            return _flatten(self.inner(*args))
+
+        in_vars, in_struct = _flatten(args, self.state_dict(keep_vars=True).values())
+        trace, (out_vars, out_struct) = traced_inner(in_vars, in_struct)
+        out, extra = _unflatten(out_vars, out_struct)
+        assert len(extra) == 0
+        return trace, out
+
+
+# Functional version that assumes that all parameters are explicitly
+# specified
+def raw_trace(nderivs=0):
+    def _raw_trace(f):
+        @functools.wraps(f)
+        def wrapper(in_vars, in_struct=None):
+            trace = torch._C._tracer_enter(in_vars, nderivs)
+            out_vars, out_struct = f(in_vars, in_struct)
+            torch._C._tracer_exit(out_vars)
+            return trace, (out_vars, out_struct)
+        return wrapper
+    return _raw_trace
+
+
+# Lifecycle of a CompiledModule:
+# - It is given an underlying function, which knows how to actually
+#   execute the code that we want to compile.
+# - When we encounter an input configuration for which we don't
+#   have an optimized trace, we run the underlying function, tracing its
+#   result.  The trace is not done yet, so we save it into our set of pending
+#   traces for that configuration.
+# - When we encounter an input configuration whose trace is "ready"
+#   (that is, we've seen all of the passes, so the trace contains
+#   forwards/backwards/etc), we compile it, and then register this
+#   as the compiled trace.
+# - When we encounter an input configuration whose trace is compiled,
+#   we just directly run the compiled trace.
+class CompiledModule(Module):
+    def __init__(self, inner, params=tuple(), **kwargs):
+        super(CompiledModule, self).__init__()
+        self.inner = inner
+        self.params = ParameterList(list(params))
+        self.kwargs = kwargs
+        self.ktrace_cache = {}
+
+    def _process_args(self, args):
+        in_vars, in_struct = _flatten(args, self.state_dict(keep_vars=True).values())
+        # TODO: I'm not entirely sure about volatile implies nderivs=0 special case
+        is_volatile, in_vars_key = vars_key(in_vars)
+        in_key = (in_vars_key, in_struct)
+        return in_vars, in_struct, is_volatile, in_key
+
+    # NB: In principle, there could also be a 'raw' version of this compiler,
+    # but since the logic is so complicated, testing code wouldn't benefit much
+    def forward(self, *args):
+        in_vars, in_struct, is_volatile, in_key = self._process_args(args)
+        ktrace = self.ktrace_cache.get(in_key)
+        if ktrace is None:
+            ktrace = TraceForKey(self.inner, volatile=is_volatile, **self.kwargs)
+            self.ktrace_cache[in_key] = ktrace
+        # TODO: avoid virtual method call here
```

Comment:
lol. Every attribute access means you have to do like 3 virtual method calls

Reply:
I wasn't sure :o)